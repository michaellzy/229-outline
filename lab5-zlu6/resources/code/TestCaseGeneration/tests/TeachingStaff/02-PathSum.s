# $s0: N
# $s1: ID
# $s2: N_ID

PathSum:
    add  $sp, $sp, -16
    sw   $s0, 0($sp)
    sw   $s1, 4($sp)
    sw   $s2, 8($sp)
    sw   $ra, 12($sp)
    bne  $a0, $zero, check_id # if N != NULL

    addi $v0, $0, -1          # $v0 <-- -1
    j    done
	
check_id:
    lw   $s2, 0($a0)          # N_ID <-- N->ID
    bne  $s2, $a1, go_left    # if N_ID != ID
	
    add  $v0, $0, $a1         # $v0 <-- ID
    j    done
	
go_left:
    add  $s1, $0, $a1         # $s1 <-- ID
    add  $s0, $0, $a0         # $s0 <-- N
    lw   $a0, 4($s0)          # $a0 <-- N->left
    jal  PathSum              # PathSum(N->left, ID)
	
    addi $t1, $0, -1
    bne  $v0, $t1 return      # if $v0 != -1
	
    lw   $a0, 8($s0)          # $a0 <-- N->right
    add  $a1, $0, $s1         # $a1 <-- ID
    jal  PathSum              # PathSum(N->right, ID)
	
    addi $t1, $0, -1
    beq  $v0, $t1, done       # if $v0 == -1
	
return:
    add  $v0, $v0, $s2        # $v0 <-- found + N_ID
	
done:
    lw   $s0, 0($sp)
    lw   $s1, 4($sp)
    lw   $s2, 8($sp)
    lw   $ra, 12($sp)
    addi  $sp, $sp, 16
    jr  $ra
