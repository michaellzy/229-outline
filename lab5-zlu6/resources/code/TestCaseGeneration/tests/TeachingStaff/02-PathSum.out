10 block(s) found.
Block Leader: 0x10010000, Size: 6
Block Leader: 0x10010018, Size: 2
Block Leader: 0x10010020, Size: 2
Block Leader: 0x10010028, Size: 2
Block Leader: 0x10010030, Size: 4
Block Leader: 0x10010040, Size: 2
Block Leader: 0x10010048, Size: 3
Block Leader: 0x10010054, Size: 2
Block Leader: 0x1001005C, Size: 1
Block Leader: 0x10010060, Size: 6

Edges:
0x10010000 --> 0x10010018
0x10010000 --> 0x10010020
0x10010018 --> 0x10010060
0x10010020 --> 0x10010028
0x10010020 --> 0x10010030
0x10010028 --> 0x10010060
0x10010030 --> 0x10010040
0x10010040 --> 0x10010048
0x10010040 --> 0x1001005C
0x10010048 --> 0x10010054
0x10010054 --> 0x1001005C
0x10010054 --> 0x10010060
0x1001005C --> 0x10010060

Dominator Bit Vectors:
0000 0000 0000 0000 0000 0000 0000 0001 
0000 0000 0000 0000 0000 0000 0000 0011 
0000 0000 0000 0000 0000 0000 0000 0101 
0000 0000 0000 0000 0000 0000 0000 1101 
0000 0000 0000 0000 0000 0000 0001 0101 
0000 0000 0000 0000 0000 0000 0011 0101 
0000 0000 0000 0000 0000 0000 0111 0101 
0000 0000 0000 0000 0000 0000 1111 0101 
0000 0000 0000 0000 0000 0001 0011 0101 
0000 0000 0000 0000 0000 0010 0000 0001 
